feat: add server dataset hub import

jfcalvo opened the pull request:
# Description

This PR add changes to support the import of datasets from HF Hub into Argilla with the following features:
* Add a new `POST /api/v1/datasets/:dataset_id/import` endpoint that will enqueue a background job to import records from a HF Hub dataset, returning information about the enqueued job. It expect the following parameters:
  * `name`: the name of the dataset (i.e. `lhoestq/demo1`) @burtenshaw suggested changing it to `repo_id`
  * `subset`: the dataset subset (i.e. `default`) @burtenshaw suggested to make the parameter optional
  * `split`: the dataset split (i.e. `train`) @burtenshaw suggested to make the parameter opcional
* Add a new background job so the import process can be done outside of request time.
* Add a new `HubDataset` class encapsulating all the logic to import a dataset from the Hub.
* Add a new `/api/v1/jobs/:job_id` to get information about the status of one specific job. This is useful if the UI or the SDK needs to know if the import process finished. (@frascuchon we can use this to give information about other processes, for example when a dataset distribution settings is changed).

Refs https://github.com/argilla-io/roadmap/issues/21

**Type of change**

- New feature (non-breaking change which adds functionality)

**How Has This Been Tested**

- [x] Manually testing and adding more automatic tests to our suite.

**Checklist**

- I added relevant documentation
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature works
- I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

jfcalvo committed:
feat: first iteration of background job to import datasets from hub

jfcalvo commented on argilla-server/src/argilla_server/jobs/hub_jobs.py:
```diff
@@ -0,0 +1,48 @@
+# TODO: Once we merge webhooks we should change the queue to use a different one (default queue is deleted there)
+@job(DEFAULT_QUEUE, timeout=JOB_TIMEOUT_DISABLED, retry=Retry(max=3))
+async def import_dataset_from_hub_job(name: str, subset: str, split: str, dataset_id: UUID) -> None:
+    hub_dataset = HubDataset(name, subset, split)
```
Move this after loading dataset from database.

jfcalvo committed:
feat: improve import_dataset_from_hub_job to get dataset before instantiate HubDataset class

jfcalvo committed:
feat: improve HubDataset batch processing

burtenshaw commented on argilla-server/pyproject.toml:
```diff
@@ -59,6 +59,7 @@ dependencies = [
     "typer >= 0.6.0, < 0.10.0", # spaCy only supports typer<0.10.0
     "packaging>=23.2",
     "psycopg2-binary>=2.9.9",
+    "datasets>=3.0.1",
```
I think this is a bit strict. That said, it's the server so compatibility is less important.

jfcalvo committed:
feat: use UpsertRecordsBulk of CreateRecordsBulk for importing datasets from hub

jfcalvo committed:
feat: transform dataset importing value columns with PIL images to data URLs

jfcalvo committed:
feat: add support to map suggestions importing datasets from hub

frascuchon committed:
Merge branch 'feat/argilla-direct-feature-branch' into feat/add-hub-dataset-import

frascuchon committed:
Merge branch 'feat/argilla-direct-feature-branch' into feat/add-hub-dataset-import

frascuchon commented on argilla-server/src/argilla_server/api/handlers/v1/datasets/datasets.py:
```diff
@@ -301,3 +304,26 @@ async def update_dataset(
     return await datasets.update_dataset(db, dataset, dataset_update.dict(exclude_unset=True))
+
+
+# TODO: Maybe change /import to /import-from-hub?
```
Why not reflect the import source as part of the body? We could reuse the same import endpoint to import data from other sources.

jfcalvo replied:
Yes, that's another possibility.

frascuchon commented on argilla-server/src/argilla_server/jobs/hub_jobs.py:
```diff
@@ -0,0 +1,47 @@
+        )
+
+        async with SearchEngine.get_by_name(settings.search_engine) as search_engine:
+            await HubDataset(name, subset, split).import_to(db, search_engine, dataset)
```
WDYT?
```suggestion
            await HubDataset(name, subset, split).export_to(db, search_engine, dataset)
```

jfcalvo committed:
feat: add support for hub dataset mapping

jfcalvo committed:
feat: set metadata and suggestions as optional for HubDatasetMapping

jfcalvo committed:
feat: when no external_id is mapped row_idx is used

jfcalvo committed:
feat: use streaming when loading the dataset

jfcalvo committed:
feat: refactor UpsertRecordsBulk to validate records individually

jfcalvo committed:
feat: ignore invalid records when importing datasets from hub

frascuchon commented on argilla-server/src/argilla_server/bulk/records_bulk.py:
```diff
@@ -207,7 +215,13 @@ async def upsert_records_bulk(self, dataset: Dataset, bulk_upsert: RecordsBulkUp
+                            ) from ex
+                        else:
+                            # NOTE: Ignore the errors in this record and continue with the next one
+                            continue
```
Logging here could help

jfcalvo committed:
Merge branch 'feat/argilla-direct-feature-branch' into feat/add-hub-dataset-import

frascuchon committed:
Merge branch 'feat/argilla-direct-feature-branch' into feat/add-hub-dataset-import

jfcalvo committed:
feat: add a fixed number of rows to take importing dataset from Hub (#5597)

# Description

I have added a number of row to take importing a dataset from the hub.
Specifically 500K rows. This can help us to avoid importing really big
datasets with millions of rows into Argilla.

Refs https://github.com/argilla-io/roadmap/issues/21

**Type of change**

- New feature (non-breaking change which adds functionality)

**How Has This Been Tested**

- [ ] Manually tested importing a dataset with a big number of records
(> 500K)

**Checklist**

- I added relevant documentation
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature
works
- I have added relevant notes to the CHANGELOG.md file (See
https://keepachangelog.com/)

---------

Co-authored-by: Paco Aranda <frascuchon@gmail.com>

jfcalvo committed:
Merge branch 'feat/argilla-direct-feature-branch' into feat/add-hub-dataset-import

jfcalvo committed:
feat: add support for class labels and casting rows (#5601)

# Description

This PR adds the following features:
* Add support to dataset `ClassLabel` features casting them using
`int2str` so we store string values on the Argilla imported dataset.
* Now the casting is done at the row level and discarding keys that are
not part of the mapping sources. Once the values arrive to the different
record create values these are already casted.
* We are casting `ClassLabel` features to string and `Image` features to
data-url strings.

Refs https://github.com/argilla-io/roadmap/issues/21

**Type of change**

- New feature (non-breaking change which adds functionality)

**How Has This Been Tested**

- [x] Adding additional tests import real datasets from HF.

**Checklist**

- I added relevant documentation
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature
works
- I have added relevant notes to the CHANGELOG.md file (See
https://keepachangelog.com/)

jfcalvo committed:
feat: improve `HubDataset` image processing support (#5606)

# Description

This PR adds the following changes related to how `HubDataset` import
functionality process rows with images:
* If the image has not format we transform the image to `png`.
* We convert images to `RGB` color space to avoid problems with other
unsupported color spaces.

Refs https://github.com/argilla-io/roadmap/issues/21

**Type of change**

- New feature (non-breaking change which adds functionality)

**How Has This Been Tested**

- [x] Manually testing `microsoft/cats_vs_dogs` dataset.

**Checklist**

- I added relevant documentation
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature
works
- I have added relevant notes to the CHANGELOG.md file (See
https://keepachangelog.com/)

jfcalvo committed:
feat: add support to `-1` no label values for `ClassLabel` dataset features (#5607)

# Description

This PR adds the following changes:
* Add support to `ClassLabel` features using `-1` (no label) values and
casting these values to be `None`.
* Now when importing `fields`, `metadata`, or `suggestions` and the
value is `None` they will be ignored.

Refs https://github.com/argilla-io/roadmap/issues/21

**Type of change**

- New feature (non-breaking change which adds functionality)

**How Has This Been Tested**

- [x] Adding new tests and modifying existing ones.

**Checklist**

- I added relevant documentation
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature
works
- I have added relevant notes to the CHANGELOG.md file (See
https://keepachangelog.com/)

---------

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
